### PR TITLE
feat(core): password checking api

### PR DESCRIPTION
**Pull request by gao-sun**
## Summary

Add `POST /api/sign-in-exp/default/check-password` API to check if the password meets the password policy configured in the default sign-in experience. See changeset for details.

## Testing

A lot of integration tests added.

## Checklist

- [x] `.changeset`
- [ ] unit tests
- [x] integration tests
- [x] API documented

### CODE REVIEW

File: packages/core/src/routes-me/user.ts [hunk 1]
```diff
@@ -129,16 +130,27 @@ export default function userRoutes<T extends AuthedMeRouter>(
...
 
-      assertThat(!isSuspended, new RequestError({ code: 'user.suspended', status: 401 }));
+      assertThat(!user.isSuspended, new RequestError({ code: 'user.suspended', status: 401 }));
 
-      await checkVerificationStatus(userId, null);
+      const [signInExperience] = await Promise.all([
+        findDefaultSignInExperience(),
+        checkVerificationStatus(userId, null),
+      ]);
+      const passwordPolicyChecker = new PasswordPolicyChecker(signInExperience.passwordPolicy);
```
Comment (charIeszhao) [inline packages/core/src/routes-me/user.ts:146]: I think we should also pass user to the policy checker

File: packages/core/src/routes-me/user.ts [hunk 2]
```diff
@@ -129,16 +130,27 @@ export default function userRoutes<T extends AuthedMeRouter>(
...
+      const issues = await checkPasswordPolicyForUser(passwordPolicyChecker, password, user);
+
+      if (issues.length > 0) {
+        ctx.status = 400;
+        ctx.body = { issues };
+        return next();
+      }
```
Comment (charIeszhao) [inline packages/core/src/routes-me/user.ts:153]: Why don't we throw RequestError? The frontend needs to handle the error and display error messages

File: packages/core/src/routes/sign-in-experience/index.ts [hunk 3]
```diff
@@ -125,5 +130,51 @@ export default function signInExperiencesRoutes<T extends ManagementApiRouter>(
...
+      const issues = await tryThat(
+        async () =>
+          user
+            ? checkPasswordPolicyForUser(passwordPolicyChecker, password, user)
+            : passwordPolicyChecker.check(password),
+        (error) => {
+          if (error instanceof TypeError) {
+            throw new RequestError('request.invalid_input', { message: error.message });
+          }
+          throw error;
```
Comment (charIeszhao) [inline packages/core/src/routes/sign-in-experience/index.ts:164]: I think the behavior is a bit different from https://github.com/logto-io/logto/blob/master/packages/core/src/routes/interaction/utils/validate-password.ts#L68-L75
Should we keep them consistent?

File: packages/core/src/routes-me/user.ts (hunk 1 above)
Comment (gao-sun) [reply packages/core/src/routes-me/user.ts:146]: how?

File: packages/core/src/routes/sign-in-experience/index.ts (hunk 3 above)
Comment (gao-sun) [reply packages/core/src/routes/sign-in-experience/index.ts:164]: no it's not the same scenario. can you check changeset?

File: packages/core/src/routes-me/user.ts (hunk 1 above)
Comment (gao-sun) [reply packages/core/src/routes-me/user.ts:146]: this constructor has no user parameter, and the user is passed right after this line. so how can i pass user here and why should i do it?

File: packages/core/src/routes-me/user.ts [hunk 4]
```diff
@@ -129,16 +130,27 @@ export default function userRoutes<T extends AuthedMeRouter>(
...
+    koaGuard({ body: object({ password: string().min(1) }), status: [204, 400] }),
     async (ctx, next) => {
       const { id: userId } = ctx.auth;
       const { password } = ctx.guard.body;
 
-      const { isSuspended } = await findUserById(userId);
+      const user = await findUserById(userId);
 
-      assertThat(!isSuspended, new RequestError({ code: 'user.suspended', status: 401 }));
+      assertThat(!user.isSuspended, new RequestError({ code: 'user.suspended', status: 401 }));
```
Comment (darcyYe) [inline packages/core/src/routes-me/user.ts:140]: should also include 401 to response status guard?

File: packages/core/src/routes-me/user.ts (hunk 1 above)
Comment (charIeszhao) [reply packages/core/src/routes-me/user.ts:146]: Oh yes, the user is not passed through the constructor, but the `.check()` function. My bad.

File: packages/core/src/routes/sign-in-experience/index.ts (hunk 3 above)
Comment (charIeszhao) [reply packages/core/src/routes/sign-in-experience/index.ts:164]: Ah, I see. It makes no sense that the management API should throw in this scenario. OK

File: packages/core/src/routes/sign-in-experience/index.openapi.json [hunk 5]
```diff
@@ -132,6 +132,36 @@
           }
         }
       }
+    },
+    "/api/sign-in-exp/default/check-password": {
```
Comment (simeng-li) [inline packages/core/src/routes/sign-in-experience/index.openapi.json:136]: Need a custom operation name. The openapi endpoint integration tests failed. Can't generate the operation name automatically based on the path.

**Review by darcyYe — APPROVED** [review]

**Review by simeng-li — APPROVED** [review]
LGTM

**Review by charIeszhao — APPROVED** [review]

**Pull request merged**